File: src/cpp/packedrtree.cpp
#include "packedrtree.h"

namespace FlatGeobuf
{

void Node::expand(const Node& r)
{
    if (r.minX < minX) minX = r.minX;
    if (r.minY < minY) minY = r.minY;
    if (r.maxX > maxX) maxX = r.maxX;
    if (r.maxY > maxY) maxY = r.maxY;
}

Node Node::create(uint64_t index)
{
    return {
        std::numeric_limits<double>::infinity(),
        std::numeric_limits<double>::infinity(),
        -1 * std::numeric_limits<double>::infinity(),
        -1 * std::numeric_limits<double>::infinity(),
        index
    };
}

bool Node::intersects(const Node& r) const
{
    if (maxX < r.minX) return false;
    if (maxY < r.minY) return false;
    if (minX > r.maxX) return false;
    if (minY > r.maxY) return false;
    return true;
}

std::vector<double> Node::toVector()
{
    return std::vector<double> { minX, minY, maxX, maxY };
}

// Based on public domain code at https://github.com/rawrunprotected/hilbert_curves
uint32_t hilbert(uint32_t x, uint32_t y)
{
    uint32_t a = x ^ y;
    uint32_t b = 0xFFFF ^ a;
    uint32_t c = 0xFFFF ^ (x | y);
    uint32_t d = x & (y ^ 0xFFFF);

    uint32_t A = a | (b >> 1);
    uint32_t B = (a >> 1) ^ a;
    uint32_t C = ((c >> 1) ^ (b & (d >> 1))) ^ c;
    uint32_t D = ((a & (c >> 1)) ^ (d >> 1)) ^ d;

    a = A; b = B; c = C; d = D;
    A = ((a & (a >> 2)) ^ (b & (b >> 2)));
    B = ((a & (b >> 2)) ^ (b & ((a ^ b) >> 2)));
    C ^= ((a & (c >> 2)) ^ (b & (d >> 2)));
    D ^= ((b & (c >> 2)) ^ ((a ^ b) & (d >> 2)));

    a = A; b = B; c = C; d = D;
    A = ((a & (a >> 4)) ^ (b & (b >> 4)));
    B = ((a & (b >> 4)) ^ (b & ((a ^ b) >> 4)));
    C ^= ((a & (c >> 4)) ^ (b & (d >> 4)));
    D ^= ((b & (c >> 4)) ^ ((a ^ b) & (d >> 4)));

    a = A; b = B; c = C; d = D;
    C ^= ((a & (c >> 8)) ^ (b & (d >> 8)));
    D ^= ((b & (c >> 8)) ^ ((a ^ b) & (d >> 8)));

    a = C ^ (C >> 1);
    b = D ^ (D >> 1);

    uint32_t i0 = x ^ y;
    uint32_t i1 = b | (0xFFFF ^ (i0 | a));

    i0 = (i0 | (i0 << 8)) & 0x00FF00FF;
    i0 = (i0 | (i0 << 4)) & 0x0F0F0F0F;
    i0 = (i0 | (i0 << 2)) & 0x33333333;
    i0 = (i0 | (i0 << 1)) & 0x55555555;

    i1 = (i1 | (i1 << 8)) & 0x00FF00FF;
    i1 = (i1 | (i1 << 4)) & 0x0F0F0F0F;
    i1 = (i1 | (i1 << 2)) & 0x33333333;
    i1 = (i1 | (i1 << 1)) & 0x55555555;

    uint32_t value = ((i1 << 1) | i0);

    return value;
}

uint32_t hilbert(const Node& r, uint32_t hilbertMax, const Node& extent)
{
    uint32_t x = static_cast<uint32_t>(floor(hilbertMax * ((r.minX + r.maxX) / 2 - extent.minX) / extent.width()));
    uint32_t y = static_cast<uint32_t>(floor(hilbertMax * ((r.minY + r.maxY) / 2 - extent.minY) / extent.height()));
    uint32_t v = hilbert(x, y);
    return v;
}

const uint32_t hilbertMax = (1 << 16) - 1;

void hilbertSort(std::vector<std::shared_ptr<Item>> &items)
{
    Node extent = std::accumulate(items.begin(), items.end(), Node::create(0), [] (Node a, std::shared_ptr<Item> b) {
        a.expand(b->node);
        return a;
    });
    std::sort(items.begin(), items.end(), [&extent] (std::shared_ptr<Item> a, std::shared_ptr<Item> b) {
        uint32_t ha = hilbert(a->node, hilbertMax, extent);
        uint32_t hb = hilbert(b->node, hilbertMax, extent);
        return ha > hb;
    });
}

Node calcExtent(const std::vector<Node> &nodes)
{
    Node extent = std::accumulate(nodes.begin(), nodes.end(), Node::create(0), [] (Node a, const Node& b) {
        a.expand(b);
        return a;
    });
    return extent;
}

Node calcExtent(const std::vector<std::shared_ptr<Item>> &items)
{
    Node extent = std::accumulate(items.begin(), items.end(), Node::create(0), [] (Node a, std::shared_ptr<Item> b) {
        a.expand(b->node);
        return a;
    });
    return extent;
}

void hilbertSort(std::vector<Node> &items)
{
    Node extent = calcExtent(items);
    std::sort(items.begin(), items.end(), [&extent] (const Node& a, const Node& b) {
        uint32_t ha = hilbert(a, hilbertMax, extent);
        uint32_t hb = hilbert(b, hilbertMax, extent);
        return ha > hb;
    });
}

void PackedRTree::init(const uint16_t nodeSize)
{
    if (nodeSize < 2)
        throw std::invalid_argument("Node size must be at least 2");
    if (_numItems == 0)
        throw std::invalid_argument("Cannot create empty tree");
    _nodeSize = std::min(std::max(nodeSize, static_cast<uint16_t>(2)), static_cast<uint16_t>(65535));
    _levelBounds = generateLevelBounds(_numItems, _nodeSize);
    _numNodes = _levelBounds.back();
    _nodes.reserve(static_cast<size_t>(_numNodes));
}

std::vector<uint64_t> PackedRTree::generateLevelBounds(const uint64_t numItems, const uint16_t nodeSize) {
    if (nodeSize < 2)
        throw std::invalid_argument("Node size must be at least 2");
    if (numItems == 0)
        throw std::invalid_argument("Number of items must be greater than 0");
    if (numItems > std::numeric_limits<uint64_t>::max() - ((numItems / nodeSize) * 2))
        throw std::overflow_error("Number of items too large");
    std::vector<uint64_t> levelBounds;
    uint64_t n = numItems;
    uint64_t numNodes = n;
    levelBounds.push_back(n);
    do {
        n = (n + nodeSize - 1) / nodeSize;
        numNodes += n;
        levelBounds.push_back(numNodes);
    } while (n != 1);
    return levelBounds;
}

void PackedRTree::generateNodes()
{
    for (uint32_t i = 0, pos = 0; i < _levelBounds.size() - 1; i++) {
        uint32_t end = static_cast<uint32_t>(_levelBounds[i]);
        while (pos < end) {
            Node node = Node::create(pos);
            for (uint32_t j = 0; j < _nodeSize && pos < end; j++)
                node.expand(_nodes[pos++]);
            _nodes.push_back(node);
        }
    }
}

void PackedRTree::fromData(const void *data)
{
    auto buf = reinterpret_cast<const uint8_t *>(data);
    const Node *pn = reinterpret_cast<const Node *>(buf);
    for (uint64_t i = 0; i < _numNodes; i++) {
        Node n = *pn++;
        _nodes.push_back(n);
        _extent.expand(n);
    }
}

static std::vector<Node> convert(const std::vector<std::shared_ptr<Item>> &items)
{
    std::vector<Node> nodes;
    for (const std::shared_ptr<Item> item: items)
        nodes.push_back(item->node);
    return nodes;
}

PackedRTree::PackedRTree(const std::vector<std::shared_ptr<Item>> &items, const Node& extent, const uint16_t nodeSize) :
    _extent(extent),
    _nodes(convert(items)),
    _numItems(items.size())
{
    init(nodeSize);
    generateNodes();
}

PackedRTree::PackedRTree(const std::vector<Node> &nodes, const Node& extent, const uint16_t nodeSize) :
    _extent(extent),
    _nodes(nodes),
    _numItems(nodes.size())
{
    init(nodeSize);
    generateNodes();
}

PackedRTree::PackedRTree(const void *data, const uint64_t numItems, const uint16_t nodeSize) :
    _extent(Node::create(0)),
    _numItems(numItems)
{
    init(nodeSize);
    fromData(data);
}

std::vector<uint64_t> PackedRTree::search(double minX, double minY, double maxX, double maxY) const
{
    Node n { minX, minY, maxX, maxY };
    std::vector<uint64_t> queue;
    std::vector<uint64_t> results;
    queue.push_back(_nodes.size() - 1);
    queue.push_back(_levelBounds.size() - 1);
    while(queue.size() != 0) {
        uint64_t nodeIndex = queue[queue.size() - 2];
        uint64_t level = queue[queue.size() - 1];
        queue.pop_back();
        queue.pop_back();
        // find the end index of the node
        uint64_t end = std::min(static_cast<uint64_t>(nodeIndex + _nodeSize), _levelBounds[static_cast<size_t>(level)]);
        // search through child nodes
        for (uint64_t pos = nodeIndex; pos < end; pos++) {
            auto node = _nodes[static_cast<size_t>(pos)];
            if (!n.intersects(node))
                continue;
            if (nodeIndex < _numItems) {
                results.push_back(pos); // leaf item
            } else {
                queue.push_back(node.index); // node; add it to the search queue
                queue.push_back(level - 1);
            }
        }
    }
    return results;
}

std::vector<uint64_t> PackedRTree::streamSearch(
    const uint64_t numItems, const uint16_t nodeSize, const Node& n,
    const std::function<void(uint8_t *, size_t, size_t)> &readNode)
{
    auto levelBounds = generateLevelBounds(numItems, nodeSize);
    uint64_t numNodes = levelBounds.back();
    std::vector<Node> nodes;
    nodes.reserve(nodeSize);
    uint8_t *nodesBuf = reinterpret_cast<uint8_t *>(nodes.data());
    std::vector<uint64_t> queue;
    std::vector<uint64_t> results;
    queue.push_back(numNodes - 1);
    queue.push_back(levelBounds.size() - 1);
    while(queue.size() != 0) {
        uint64_t nodeIndex = queue[queue.size() - 2];
        bool isLeafNode = nodeIndex < numItems;
        uint64_t level = queue[queue.size() - 1];
        queue.pop_back();
        queue.pop_back();
        // find the end index of the node
        uint64_t end = std::min(static_cast<uint64_t>(nodeIndex + nodeSize), levelBounds[static_cast<size_t>(level)]);
        uint64_t length = end - nodeIndex;
        readNode(nodesBuf, static_cast<size_t>(nodeIndex * sizeof(Node)), static_cast<size_t>(length * sizeof(Node)));
        // search through child nodes
        for (uint64_t pos = nodeIndex; pos < end; pos++) {
            uint64_t nodePos = pos - nodeIndex;
            auto node = nodes[static_cast<size_t>(nodePos)];
            if (!n.intersects(node))
                continue;
            if (isLeafNode) {
                results.push_back(pos); // leaf item
            } else {
                queue.push_back(node.index); // node; add it to the search queue
                queue.push_back(level - 1);
            }
        }
    }
    return results;
}

uint64_t PackedRTree::size() const { return _numNodes * sizeof(Node); }

uint64_t PackedRTree::size(const uint64_t numItems, const uint16_t nodeSize)
{
    if (nodeSize < 2)
        throw std::invalid_argument("Node size must be at least 2");
    if (numItems == 0)
        throw std::invalid_argument("Number of items must be greater than 0");
    const uint16_t nodeSizeMin = std::min(std::max(nodeSize, static_cast<uint16_t>(2)), static_cast<uint16_t>(65535));
    // limit so that resulting size in bytes can be represented by uint64_t
    if (numItems > static_cast<uint64_t>(1) << 56)
        throw std::overflow_error("Number of items must be less than 2^56");
    uint64_t n = numItems;
    uint64_t numNodes = n;
    do {
        n = (n + nodeSizeMin - 1) / nodeSizeMin;
        numNodes += n;
    } while (n != 1);
    return numNodes * sizeof(Node);
}

void PackedRTree::streamWrite(const std::function<void(uint8_t *, size_t)> &writeData) {
    writeData(reinterpret_cast<uint8_t *>(_nodes.data()), _nodes.size() * sizeof(Node));
}

Node PackedRTree::getExtent() const { return _extent; }

}

File: src/ts/packedrtree.ts
import Config from './config.js';
import Logger from './logger.js';

export const NODE_ITEM_LEN: number = 8 * 4 + 8;
// default branching factor of a node in the rtree
//
// actual value will be specified in the header but
// this can be useful for having reasonably sized guesses for fetch-sizes when
// streaming results
export const DEFAULT_NODE_SIZE = 16;

export interface Rect {
    minX: number;
    minY: number;
    maxX: number;
    maxY: number;
}

export function calcTreeSize(numItems: number, nodeSize: number): number {
    nodeSize = Math.min(Math.max(+nodeSize, 2), 65535);
    let n = numItems;
    let numNodes = n;
    do {
        n = Math.ceil(n / nodeSize);
        numNodes += n;
    } while (n !== 1);
    return numNodes * NODE_ITEM_LEN;
}

/**
 * returns [levelOffset, numNodes] for each level
 */
export function generateLevelBounds(
    numItems: number,
    nodeSize: number,
): Array<[number, number]> {
    if (nodeSize < 2) throw new Error('Node size must be at least 2');
    if (numItems === 0)
        throw new Error('Number of items must be greater than 0');

    // number of nodes per level in bottom-up order
    let n = numItems;
    let numNodes = n;
    const levelNumNodes = [n];
    do {
        n = Math.ceil(n / nodeSize);
        numNodes += n;
        levelNumNodes.push(n);
    } while (n !== 1);

    // bounds per level in reversed storage order (top-down)
    const levelOffsets: Array<number> = [];
    n = numNodes;
    for (const size of levelNumNodes) {
        levelOffsets.push(n - size);
        n -= size;
    }
    const levelBounds: Array<[number, number]> = [];
    for (let i = 0; i < levelNumNodes.length; i++)
        levelBounds.push([levelOffsets[i], levelOffsets[i] + levelNumNodes[i]]);
    return levelBounds;
}

type ReadNodeFn = (treeOffset: number, size: number) => Promise<ArrayBuffer>;

/**
 * A feature found to be within the bounding box `rect`
 *
 *  (offset, index)
 *  `offset`: Byte offset in feature data section
 *  `index`: feature number
 *  `featureLength`: featureLength, except for the last element
 */
export type SearchResult = [number, number, number | null];

/**
 * Yield's a `SearchResult` for each feature within the bounds of `rect`.
 *
 * Every node in the FGB index tree has a bounding rect, all of the nodes children
 * are contained within that bounding rect. The leaf nodes of the tree represent
 * the features of the collection.
 *
 * As we traverse the tree, starting from the root, we'll need to read more data
 * from the index. When we don't already have this range data buffered locally,
 * an HTTP fetch is triggered. For performance, we merge adjacent and nearby
 * request ranges into a single request, reasoning that fetching a few extra
 * bytes is a good tradeoff if it means we can reduce the number of requests.
 */
export async function* streamSearch(
    numItems: number,
    nodeSize: number,
    rect: Rect,
    readNode: ReadNodeFn,
): AsyncGenerator<SearchResult, void, unknown> {
    class NodeRange {
        _level: number;
        nodes: [number, number];
        constructor(nodes: [number, number], level: number) {
            this._level = level;
            this.nodes = nodes;
        }

        level(): number {
            return this._level;
        }

        startNode(): number {
            return this.nodes[0];
        }

        endNode(): number {
            return this.nodes[1];
        }

        extendEndNodeToNewOffset(newOffset: number) {
            console.assert(newOffset > this.nodes[1]);
            this.nodes[1] = newOffset;
        }

        toString(): string {
            return `[NodeRange level: ${this._level}, nodes: ${this.nodes[0]}-${this.nodes[1]}]`;
        }
    }

    const { minX, minY, maxX, maxY } = rect;
    Logger.info(`tree items: ${numItems}, nodeSize: ${nodeSize}`);
    const levelBounds = generateLevelBounds(numItems, nodeSize);
    const leafNodesOffset = levelBounds[0][0];

    const rootNodeRange: NodeRange = (() => {
        const range: [number, number] = [0, 1];
        const level = levelBounds.length - 1;
        return new NodeRange(range, level);
    })();

    const queue: Array<NodeRange> = [rootNodeRange];

    Logger.debug(
        `starting stream search with queue: ${queue}, numItems: ${numItems}, nodeSize: ${nodeSize}, levelBounds: ${levelBounds}`,
    );

    while (queue.length != 0) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const nodeRange = queue.shift()!;

        Logger.debug(`popped node: ${nodeRange}, queueLength: ${queue.length}`);

        const nodeIndex = nodeRange.startNode();
        const isLeafNode = nodeIndex >= leafNodesOffset;

        // find the end index of the node
        const [, levelBound] = levelBounds[nodeRange.level()];

        const end = Math.min(nodeRange.endNode() + nodeSize, levelBound);
        const length = end - nodeIndex;

        const buffer = await readNode(
            nodeIndex * NODE_ITEM_LEN,
            length * NODE_ITEM_LEN,
        );

        const dataView = new DataView(buffer);
        for (let pos = nodeIndex; pos < end; pos++) {
            const nodePos = (pos - nodeIndex) * 40;
            if (maxX < dataView.getFloat64(nodePos + 0, true)) continue; // maxX < nodeMinX
            if (maxY < dataView.getFloat64(nodePos + 8, true)) continue; // maxY < nodeMinY
            if (minX > dataView.getFloat64(nodePos + 16, true)) continue; // minX > nodeMaxX
            if (minY > dataView.getFloat64(nodePos + 24, true)) continue; // minY > nodeMaxY

            const offset = dataView.getBigUint64(nodePos + 32, true);

            if (isLeafNode) {
                const featureLength = (() => {
                    if (pos < numItems - 1) {
                        // Since features are tightly packed, we infer the
                        // length of _this_ feature by measuring to the _next_
                        // feature's start.
                        const nextPos = (pos - nodeIndex + 1) * 40;
                        const nextOffset = dataView.getBigUint64(
                            nextPos + 32,
                            true,
                        );
                        return nextOffset - offset;
                    } else {
                        // This is the last feature - there's no "next" feature
                        // to measure to, so we can't know it's length.
                        return null;
                    }
                })();

                // Logger.debug(`offset: ${offset}, pos: ${pos}, featureLength: ${featureLength}`);
                yield [
                    Number(offset),
                    pos - leafNodesOffset,
                    Number(featureLength),
                ];
                continue;
            }

            // request up to this many nodes if it means we can eliminate an
            // extra request
            const extraRequestThresholdNodes =
                Config.global.extraRequestThreshold() / NODE_ITEM_LEN;

            // Since we're traversing the tree by monotonically increasing byte
            // offset, the most recently enqueued node range will be the
            // nearest, and thus presents the best candidate for merging.
            const nearestNodeRange = queue[queue.length - 1];
            if (
                nearestNodeRange !== undefined &&
                nearestNodeRange.level() == nodeRange.level() - 1 &&
                offset < nearestNodeRange.endNode() + extraRequestThresholdNodes
            ) {
                Logger.debug(
                    `Merging "nodeRange" request into existing range: ${nearestNodeRange}, newOffset: ${nearestNodeRange.endNode()} -> ${offset}`,
                );
                nearestNodeRange.extendEndNodeToNewOffset(Number(offset));
                continue;
            }

            const newNodeRange: NodeRange = (() => {
                const level = nodeRange.level() - 1;
                const range: [number, number] = [
                    Number(offset),
                    Number(offset) + 1,
                ];
                return new NodeRange(range, level);
            })();

            // We're going to add a new node range - log the reason
            if (
                nearestNodeRange !== undefined &&
                nearestNodeRange.level() == newNodeRange.level()
            ) {
                Logger.info(
                    `Same level, but too far away. Pushing new request at offset: ${offset} rather than merging with distant ${nearestNodeRange}`,
                );
            } else {
                Logger.info(
                    `Pushing new level for ${newNodeRange} onto queue with nearestNodeRange: ${nearestNodeRange} since there's not already a range for this level.`,
                );
            }

            queue.push(newNodeRange);
        }
    }
}
